core/src/Kokkos_CopyViews.hpp

```diff
@@ -1095,6 +1090,67 @@ inline void deep_copy(
   }
 }
 
+template <class DT, class... DP, class ST, class... SP>
+inline void deep_copy(
+    const View<DT, DP...>& dst, const View<ST, SP...>& src,
+    std::enable_if_t<(
+        std::is_void_v<typename ViewTraits<DT, DP...>::specialize> &&
+        std::is_void_v<typename ViewTraits<ST, SP...>::specialize> &&
+        !std::is_same_v<typename ViewTraits<DT, DP...>::value_type,
+                        typename ViewTraits<ST, SP...>::non_const_value_type> &&
+        (unsigned(ViewTraits<DT, DP...>::rank) == unsigned(0) &&
+         unsigned(ViewTraits<ST, SP...>::rank) == unsigned(0)))>* = nullptr) {
+  using dst_type = View<DT, DP...>;
+  using src_type = View<ST, SP...>;
+
+  static_assert(std::is_same_v<typename dst_type::value_type,
+                               typename dst_type::non_const_value_type>,
+                "deep_copy requires non-const destination type");
```

Comment:
Probably not required to be addressed in this PR, but worth given a thought and (probably) a dedicated issue.

In our code we've introduced the following concept:
```c++
template <typename T>
concept ModifiableView = View<T> && ! std::is_const_v<typename T::value_type>;
```
which does exactly what you're doing (check that the value type is not `const` so that we can modify the values).

I find it clearer to assert that `!std::is_const_v<typename ViewType::value_type`.

```suggestion
  static_assert(!std::is_const_v<typename dst_type::value_type>,
                "deep_copy requires non-const destination type");
```

What do you think ?

Reply:
Yeah lets looks at this in follow up. This is done in various places this way and I should look at them holistically.

Reply:
I've created #8323 for the followup :wink: